internal/rpc/otel_interceptor.go

```diff
@@ -0,0 +1,420 @@
+	"go.opentelemetry.io/otel/propagation"
+	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/TBD54566975/ftl/internal/log"
+)
+
+const (
+	otelFtlRequestKey       = "ftl.requestKey"
+	otelFtlVerbRef          = "ftl.verb.ref"
+	otelFtlVerbModule       = "ftl.verb.module"
+	otelMessageEvent        = "message"
+	otelMessageIDKey        = attribute.Key("message.id")
+	otelMessageSizeKey      = attribute.Key("message.uncompressed_size")
+	otelMessageTypeKey      = attribute.Key("message.type")
+	otelMessageTypeSent     = "SENT"
+	otelMessageTypeReceived = "RECEIVED"
+)
+
+func OtelInterceptor() connect.Interceptor {
+	return &otelInterceptor{}
+}
+
+type otelInterceptor struct{}
+
+var clientInstruments instrumentation
+var serverInstruments instrumentation
+
+func init() {
+	clientInstruments = createInstruments(otel.GetMeterProvider().Meter("ftl.rpc.client"))
+	serverInstruments = createInstruments(otel.GetMeterProvider().Meter("ftl.rpc.server"))
+}
+
+func getInstruments(isClient bool) instrumentation {
+	if isClient {
+		return clientInstruments
+	}
+	return serverInstruments
+}
+
+func getAttributes(ctx context.Context, rpcSystemKey string) []attribute.KeyValue {
+	logger := log.FromContext(ctx)
+	attributes := []attribute.KeyValue{
+		semconv.RPCSystemKey.String(rpcSystemKey),
+	}
+	requestKey, err := RequestKeyFromContext(ctx)
+	if err != nil {
+		logger.Debugf("failed to get request key: %s", err)
+	}
+	if key, ok := requestKey.Get(); ok {
+		attributes = append(attributes, attribute.String(otelFtlRequestKey, key.String()))
+	}
+	if verb, ok := VerbFromContext(ctx); ok {
+		attributes = append(attributes, attribute.String(otelFtlVerbRef, verb.String()))
+		attributes = append(attributes, attribute.String(otelFtlVerbModule, verb.Module))
+	}
+	return attributes
+}
+
+func (i *otelInterceptor) WrapUnary(next connect.UnaryFunc) connect.UnaryFunc {
+	return func(ctx context.Context, request connect.AnyRequest) (connect.AnyResponse, error) {
+		ctx = propagateOtelHeaders(ctx, request.Spec().IsClient, request.Header())
+		requestStartTime := time.Now()
+		isClient := request.Spec().IsClient
+		name := strings.TrimLeft(request.Spec().Procedure, "/")
+
+		spanKind := trace.SpanKindClient
+		requestSpan := otelMessageTypeKey.String(otelMessageTypeSent)
+		responseSpan := otelMessageTypeKey.String(otelMessageTypeReceived)
+		if !isClient {
+			spanKind = trace.SpanKindServer
+			requestSpan = otelMessageTypeKey.String(otelMessageTypeReceived)
+			responseSpan = otelMessageTypeKey.String(otelMessageTypeSent)
+		}
+
+		attributes := getAttributes(ctx, request.Peer().Protocol)
+		traceOpts := []trace.SpanStartOption{
+			trace.WithAttributes(attributes...),
+			trace.WithSpanKind(spanKind),
+		}
+		tracer := otel.GetTracerProvider().Tracer(request.Spec().Procedure)
+		ctx, span := tracer.Start(ctx, name, traceOpts...)
+		defer span.End()
+
+		var requestSize int
```

Comment:
Should this default to 0 if there's no request object? If not, we should probably mention in the description of this metric that it's only recorded for RPCs with a non-null request.